### ISSUE TITLE
Training does not seem to resume at specified 'starting_epoch'

### ISSUE THREAD

**Issue by dps42** [tensorpack/tensorpack#1175]
If you're asking about an unexpected problem which you do not know the root cause,
use this template. __PLEASE DO NOT DELETE THIS TEMPLATE, FILL IT__:

If you already know the root cause to your problem,
feel free to delete everything in this template.

### 1. What you did:

(1) **If you're using examples, what's the command you run:**

In the 'examples/basics' folder, run 'python mnist-tfslim.py'

(2) **If you're using examples, have you made any changes to the examples? Paste `git status; git diff` here:**

On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   mnist-tfslim.py

no changes added to commit (use "git add" and/or "git commit -a")

diff --git a/examples/basics/mnist-tfslim.py b/examples/basics/mnist-tfslim.py
index f340cd70..35659b3f 100755
--- a/examples/basics/mnist-tfslim.py
+++ b/examples/basics/mnist-tfslim.py
@@ -85,6 +85,7 @@ if __name__ == '__main__':
                 dataset_test,
                 ScalarStats(['cross_entropy_loss', 'accuracy'])),
         ],
-        max_epoch=100,
+        staring_epoch=10,
+        max_epoch=20,
     )
     launch_train_with_config(config, SimpleTrainer())

(3) **If not using examples, tell us what you did:**

  It's always better to copy-paste what you did than to describe them.

  Please try to provide enough information to let other __reproduce__ your issues. 
  Without reproducing the issue, we may not be able to investigate it.

### 2. What you observed:

(1) **Include the ENTIRE logs here:**

It's always better to copy-paste what you observed instead of describing them.

It's always better to paste **as much as possible**, although sometimes a partial log is OK.

Tensorpack typically saves stdout to its training log.
If stderr is relevant, you can run a command with `CMD 2>&1 | tee logs.txt`
to save both stdout and stderr to one file.

[32m[0503 09:47:36 @logger.py:90][0m Argv: mnist-tfslim.py
[32m[0503 09:47:36 @fs.py:101][0m [5m[31mWRN[0m Env var $TENSORPACK_DATASET not set, using C:\Users\dps42\tensorpack_data for datasets.
[32m[0503 09:47:36 @trainers.py:47][0m Building graph for a single training tower ...
[32m[0503 09:47:36 @regularize.py:129][0m regularize_cost_from_collection() found 2 regularizers in REGULARIZATION_LOSSES collection.
[32m[0503 09:47:37 @model_utils.py:67][0m [36mList of Trainable Variables: 
[0mname             shape             #elements
---------------  --------------  -----------
conv0/weights:0  [3, 3, 1, 32]           288
conv0/biases:0   [32]                     32
conv1/weights:0  [3, 3, 32, 32]         9216
conv1/biases:0   [32]                     32
conv2/weights:0  [3, 3, 32, 32]         9216
conv2/biases:0   [32]                     32
conv3/weights:0  [3, 3, 32, 32]         9216
conv3/biases:0   [32]                     32
fc0/weights:0    [1568, 512]          802816
fc0/biases:0     [512]                   512
fc1/weights:0    [512, 10]              5120
fc1/biases:0     [10]                     10[36m
Number of trainable variables: 12
Number of parameters (elements): 836522
Storage space needed for all trainable variables: 3.19MB[0m
[32m[0503 09:47:37 @base.py:209][0m Setup callbacks graph ...
[32m[0503 09:47:37 @inference_runner.py:149][0m [InferenceRunner] Building tower 'InferenceTower' on device /gpu:0 ...
[32m[0503 09:47:37 @summary.py:46][0m [MovingAverageSummary] 2 operations in collection 'MOVING_SUMMARY_OPS' will be run with session hooks.
[32m[0503 09:47:37 @summary.py:93][0m Summarizing collection 'summaries' of size 15.
[32m[0503 09:47:37 @base.py:230][0m Creating the session ...
[32m[0503 09:47:37 @base.py:236][0m Initializing the session ...
[32m[0503 09:47:37 @base.py:243][0m Graph Finalized.
[32m[0503 09:47:38 @inference_runner.py:96][0m [InferenceRunner] Will eval 40 iterations
[32m[0503 09:47:38 @monitor.py:352][0m [5m[31mWRN[0m History epoch=10 from JSON is not the predecessor of the current starting_epoch=10
[32m[0503 09:47:38 @monitor.py:353][0m [5m[31mWRN[0m If you want to resume old training, either use `AutoResumeTrainConfig` or correctly set the new starting_epoch yourself to avoid inconsistency. 
[32m[0503 09:47:38 @monitor.py:360][0m [5m[31mWRN[0m Now, we will train with starting_epoch=10 and backup old json to train_log\mnist-tfslim\stats.json.0503-094738
[32m[0503 09:47:38 @base.py:275][0m Start Epoch 10 ...
[32m[0503 09:47:43 @base.py:285][0m Epoch 10 (global_step 468) finished, time:4.89 seconds.
[32m[0503 09:47:43 @saver.py:79][0m Model saved to train_log\mnist-tfslim\model-468.
[32m[0503 09:47:43 @monitor.py:467][0m accuracy: 0.97723
[32m[0503 09:47:43 @monitor.py:467][0m cross_entropy_loss: 0.067853
[32m[0503 09:47:43 @monitor.py:467][0m lr: 0.001
[32m[0503 09:47:43 @monitor.py:467][0m param-summary/conv0/weights-rms: 0.10649
[32m[0503 09:47:43 @monitor.py:467][0m param-summary/conv1/weights-rms: 0.065895
[32m[0503 09:47:43 @monitor.py:467][0m param-summary/conv2/weights-rms: 0.064971
[32m[0503 09:47:43 @monitor.py:467][0m param-summary/conv3/weights-rms: 0.065108
[32m[0503 09:47:43 @monitor.py:467][0m param-summary/fc0/weights-rms: 0.030151
[32m[0503 09:47:43 @monitor.py:467][0m param-summary/fc1/weights-rms: 0.065467
[32m[0503 09:47:43 @monitor.py:467][0m validation_accuracy: 0.98604
[32m[0503 09:47:43 @monitor.py:467][0m validation_cross_entropy_loss: 0.043136
[32m[0503 09:47:43 @base.py:275][0m Start Epoch 11 ...

(2) **Other observations, if any:**
For example, CPU/GPU utilization, output images, tensorboard curves, if relevant to your issue.

### 3. What you expected, if not obvious.

If you expect higher speed, please read
http://tensorpack.readthedocs.io/tutorial/performance-tuning.html 
before posting.

If you expect certain accuracy, only in one of the two conditions can we help with it:
(1) You're unable to reproduce the accuracy documented in tensorpack examples.
(2) It appears to be a tensorpack bug.

Otherwise, how to train a model to certain accuracy is a machine learning question.
We do not answer machine learning questions and it is your responsibility to 
figure out how to make your models more accurate.
-------------------------------------------------------------------------------------
Simply, I just ran the 'examples/basics/mnist-tfslim.py'. I just wanted to see if I could resume training after stopping at some point.  To save time, I set the max_epoch=20, then stop the training with 'Ctrl + C' after the 10th epoch. Then, add 'starting_epoch=10' to 'mnist-tfslim.py' and run the same script.  But the training does not seem to resume at epoch 10, but seems to start from epoch 1.

At epoch 1, the accuracy starts somewhere in 0.97-0.98 range, and after epoch 10 the accuracy surely reaches above 0.99. But when I attempted to resume training with 'starting_epoch=10', the accuracy appeared to be in 0.97-0.98 range. 

When I tried 'AutoResumeTrainConfig', it seemed to work fine with accuracy in 0.99 range.

### 4. Your environment:
  + Paste the output of this command: `python3 -c 'import tensorpack.tfutils as u; print(u.collect_env_info())'`
    If this command failed, tell us your version of Python/TF/tensorpack.
  + You can install Tensorpack master by `pip install -U git+https://github.com/ppwwyyxx/tensorpack.git`
    and see if your issue is already solved.
  + If you're not using tensorpack under a normal command line shell (e.g.,
    using an IDE or jupyter notebook), please retry under a normal command line shell. 
  + Include relevant hardware information, e.g. number of GPUs used for training, amount of RAM.
-------------------------------------------------------------------------------------------------------
sys.platform          win32
Python                3.5.2 | packaged by conda-forge | (default, Jan 19 2017, 15:41:23) [MSC v.1900 64 bit (AMD64)]
Tensorpack            v0.9.4-46-g0c2c44e2-dirty
Numpy                 1.15.2
TensorFlow            1.4.0/b'unknown'
TF Compiler Version   MSVC
TF CUDA support       True
Nvidia Driver
CUDA
CUDNN
NCCL
CUDA_VISIBLE_DEVICES  None
GPU                   Not found with NVML
Free RAM              41.39/63.68 GB
CPU Count             8
cv2                   3.1.0
msgpack               0.6.1
python-prctl          False
--------------------------------------------------------------------------------------------
The GPU is NVIDIA QUADRO P5000. 1 GPU. GPU memory is 16GB DDR5. OS is Windows 10.
CUDA 8.0

You may often want to provide extra information related to your issue, but
at the minimum please try to provide the above information __accurately__ to save effort in the investigation.

**Comment by ppwwyyxx** [comment]
The training does start from epoch 10 as you can see from the log.
To load the model at epoch 10 you'll need to specify sess_init: https://tensorpack.readthedocs.io/tutorial/save-load.html

**Comment by dps42** [comment]
Hi. Thanks for the response. The log says that it is starting from epoch 10, but I meant that since when I stopped after epoch 10, the accuracy was above 0.99, so when I restart training from epoch 10, I should get above 0.99 accuracy too. So I meanted it appeared to have the same 'net-effect' of starting from epoch 1. Sorry for the confusion. AutoResumeTrainConfig works fine in this case, and it resumed with 0.99 accuracy just fine.

I just looked at the 'Resume Training' secton from the documentation. And the documentation suggested that I can adjust 'starting_epoch' to resume training, but it does not seem to give the desired effect. Is there an example how to restore the model with 'sess_init', and will I be able to see 0.99 accuracy for this case??

**Comment by ppwwyyxx** [comment]
You need to load the checkpoint with `session_init` in addition to setting `starting_epoch`. The documentation already explained: https://tensorpack.readthedocs.io/tutorial/save-load.html#load-a-model-to-a-session
session_init is an argument of `TrainConfig`.

**Comment by dps42** [comment]
Thank you for the clarification. With 'session_init', it works fine with the following code:
"""
    model_path = os.path.join('train_log', 'mnist-tfslim', 'model-4680')
    sess_init=SaverRestore( model_path) 

    config = TrainConfig(
        model=Model(),
        dataflow=dataset_train,
        callbacks=[
            ModelSaver(),
            InferenceRunner(
                dataset_test,
                ScalarStats(['cross_entropy_loss', 'accuracy'])),
        ],
        session_init= sess_init,
        starting_epoch=10,
        max_epoch=20,
    )
"""
I would appreciate if you could include in the tensorpack documentation that 'session_init' should be used in addition to the 'starting_epoch' for future users in 'Resume Training' section. Thank you!

**Comment by ppwwyyxx** [comment]
It is already included in the "resume training" section:

> “resume training” is mostly just “loading the last known checkpoint”. Therefore you should refer to the previous section on how to load a model.
